utils/writeable/src/utf.rs

```diff
@@ -0,0 +1,126 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use alloc::borrow::Cow;
+
+use crate::{impl_display_with_writeable, LengthHint, Writeable};
+
+use core::fmt;
+
+/// Implements [`Writeable`] for [`&[u8]`] according to the [WHATWG Encoding Standard](
+/// https://encoding.spec.whatwg.org/#utf-8-decoder).
+#[derive(Debug)]
+#[allow(clippy::exhaustive_structs)] // newtype
+pub struct PotentiallyInvalidUtf8<'a>(pub &'a [u8]);
+
+impl Writeable for PotentiallyInvalidUtf8<'_> {
+    fn write_to<W: core::fmt::Write + ?Sized>(&self, sink: &mut W) -> fmt::Result {
+        let mut remaining = self.0;
+        loop {
+            match core::str::from_utf8(remaining) {
+                Ok(valid) => {
+                    return sink.write_str(valid);
+                }
+                Err(e) => {
+                    // SAFETY: By Utf8Error invariants
+                    let valid = unsafe {
+                        core::str::from_utf8_unchecked(remaining.get_unchecked(..e.valid_up_to()))
+                    };
+                    sink.write_str(valid)?;
+                    sink.write_char(char::REPLACEMENT_CHARACTER)?;
+                    let Some(error_len) = e.error_len() else {
+                        return Ok(()); // end of string
+                    };
+                    // SAFETY: By Utf8Error invariants
+                    remaining = unsafe { remaining.get_unchecked(e.valid_up_to() + error_len..) }
+                }
+            }
+        }
+    }
+
+    fn writeable_length_hint(&self) -> crate::LengthHint {
+        // Lower bound is all valid UTF-8, upper bound is all bytes with the high bit, which become replacement characters.
+        LengthHint::between(self.0.len(), self.0.len() * 3)
+    }
+
+    fn write_to_string(&self) -> Cow<str> {
+        match core::str::from_utf8(self.0) {
+            Ok(valid) => Cow::Borrowed(valid),
+            Err(e) => {
+                // SAFETY: By Utf8Error invariants
+                let valid = unsafe {
+                    core::str::from_utf8_unchecked(self.0.get_unchecked(..e.valid_up_to()))
+                };
+
+                // Let's assume this is the only error
+                let mut out = alloc::string::String::with_capacity(
+                    self.0.len() + char::REPLACEMENT_CHARACTER.len_utf8()
+                        - e.error_len().unwrap_or(0),
+                );
+
+                out.push_str(valid);
+                out.push_str(
+                    char::REPLACEMENT_CHARACTER
+                        .encode_utf8(&mut [0; char::REPLACEMENT_CHARACTER.len_utf8()]),
+                );
```

Comment:
Suggestion: why not just use `<String as core::fmt::Write>::write_char` and `write_str` here?

Reply:
Because they're fallible

Reply:
Suggestion: use `String::push(char)` for the second one.

Reply:
Hmm, that has an extra branch which might not be optimized out. Probably cheapest to just use `push_str`

Reply:
Confirmed in godbolt that it does in fact get optimized out.